Kernel/TimerQueue.cpp

```diff
@@ -0,0 +1,71 @@
+#include <AK/SinglyLinkedList.h>
+#include <Kernel/Scheduler.h>
+#include <Kernel/TimerQueue.h>
+
+
+SinglyLinkedList<Timer>* g_timer_queue;
+
+static u64 s_next_timer_due;
+static u64 s_timer_id_count;
+void TimerQueue::initialize()
+{
+ 
+    g_timer_queue = new SinglyLinkedList<Timer>();
+}
+
+
+u64 TimerQueue::add_timer(Timer& timer)
+{
+    ASSERT(timer.expires > Scheduler::g_uptime);
+    timer.id = ++s_timer_id_count;
+    g_timer_queue->sorted_insert_slow(timer);
+    s_next_timer_due = g_timer_queue->first().expires;
+    return s_timer_id_count;
+}
+
+u64 add_timer(u64 duration, TimeUnit unit, void (callback)())
+{
+	atuo timer = Timer();
+    timer.expires = g_uptime + duration * unit;
+    timer.callback = callback;
+    return add_timer(timer);
+}
+
+void TimerQueue::sorted_insert_slow(T&& value)
+{
+	auto* new_node = new Node(move(value));
```

Comment:
Reaching directly into the `SinglyLinkedList` class here is very strange.
It would be nicer to implement this by adding an API such as `SinglyLinkedList::insert(Iterator&, T&&)`.
Then you could iterate over the list until you find your insertion point and then call `insert()` with the iterator you're using.